src/RA_AchievementSet.cpp

```diff
@@ -269,72 +236,43 @@ void AchievementSet::Reset()
     m_bProcessingActive = TRUE;
 }
 
-BOOL AchievementSet::SaveToFile()
+bool AchievementSet::SaveToFile() const
 {
-    //	Takes all achievements in this group and dumps them in the filename provided.
-    FILE* pFile = nullptr;
-    char sNextLine[2048];
-    char sMem[2048];
-    unsigned int i = 0;
+    // Commits local achievements to the file
+    auto& pLocalStorage = ra::services::ServiceLocator::GetMutable<ra::services::ILocalStorage>();
+    auto pData = pLocalStorage.WriteText(ra::services::StorageItemType::UserAchievements, std::to_wstring(static_cast<unsigned int>(g_pCurrentGameData->GetGameID())));
+    if (pData == nullptr)
+        return false;
 
-    const std::wstring sFilename = GetAchievementSetFilename(g_pCurrentGameData->GetGameID());
+    pData->WriteLine(_RA_IntegrationVersion()); // version used to create the file
+    pData->WriteLine(g_pCurrentGameData->GameTitle());
 
-    _wfopen_s(&pFile, sFilename.c_str(), L"w");
-    if (pFile != nullptr)
+    for (size_t i = 0; i < g_pLocalAchievements->NumAchievements(); ++i)
     {
-        sprintf_s(sNextLine, 2048, "0.030\n");						//	Min ver
-        fwrite(sNextLine, sizeof(char), strlen(sNextLine), pFile);
-
-        sprintf_s(sNextLine, 2048, "%s\n", g_pCurrentGameData->GameTitle().c_str());
-        fwrite(sNextLine, sizeof(char), strlen(sNextLine), pFile);
-
-        //std::vector<Achievement>::const_iterator iter = g_pActiveAchievements->m_Achievements.begin();
-        //while (iter != g_pActiveAchievements->m_Achievements.end())
-        for (i = 0; i < g_pLocalAchievements->NumAchievements(); ++i)
-        {
-            Achievement* pAch = &g_pLocalAchievements->GetAchievement(i);
-
-            ZeroMemory(sMem, 2048);
-            pAch->CreateMemString();
-
-            ZeroMemory(sNextLine, 2048);
-            sprintf_s(sNextLine, 2048, "%u:%s:%s:%s:%s:%s:%s:%s:%d:%lu:%lu:%d:%d:%s\n",
-                pAch->ID(),
-                pAch->CreateMemString().c_str(),
-                pAch->Title().c_str(),
-                pAch->Description().c_str(),
-                " ", //Ach.ProgressIndicator()=="" ? " " : Ach.ProgressIndicator(),
-                " ", //Ach.ProgressIndicatorMax()=="" ? " " : Ach.ProgressIndicatorMax(),
-                " ", //Ach.ProgressIndicatorFormat()=="" ? " " : Ach.ProgressIndicatorFormat(),
-                pAch->Author().c_str(),
-                (unsigned short)pAch->Points(),
-                (unsigned long)pAch->CreatedDate(),
-                (unsigned long)pAch->ModifiedDate(),
-                (unsigned short)pAch->Upvotes(), // Fix values
-                (unsigned short)pAch->Downvotes(), // Fix values
-                pAch->BadgeImageURI().c_str());
-
-            fwrite(sNextLine, sizeof(char), strlen(sNextLine), pFile);
-        }
-
-        fclose(pFile);
-        return TRUE;
+        Achievement* pAch = &g_pLocalAchievements->GetAchievement(i);
+
+        pData->Write(std::to_string(pAch->ID()));
+        pData->Write(":");
```

Comment:
Not a big deal now but in the future we could probably write a parameter pack overload for this function but require that every argument is a string to not require a specifier. Might not be exactly what we need but it could look like this. I did look at the Write Methods, these are just examples to show a concept, some of it might not even exist.

```
// Default:
template<typename String, typename = std::enable_if_t<is_string_v<String>>>
void Write(const String& str) noexcept { /*Do the default thing*/ }

// Variadic Version
template<typename String, typename ...Strings>
void Write(const String& str, String&&... strs) noexcept 
{ 
   static_assert(std::is_convertible_v<String, Strings>);
   for(auto i = 0U; i < sizeof...(strs); i++) // sizeof... reports the number of elements in the expansion
      Write(str, std::forward<Strings>(strs)...); // write str, forward the next strs as str   
}
```
I'm going to try something like this myself and see if I can suggest something more practical.
The intention is to have it write out like this...
`pData->Write(std::to_string(pAch->ID()), ":", pAch->CreateMemString(), ":", .... /*the other ones*/);`